pkg/service/claimlookup/claimlookup.go

```diff
@@ -1,15 +1,66 @@
 package claimlookup
 
 import (
-	"github.com/ipni/go-libipni/find/model"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	"github.com/ipfs/go-cid"
 	"github.com/storacha-network/go-ucanto/core/delegation"
+	"github.com/storacha-network/indexing-service/pkg/types"
 )
 
-// ClaimLookup is used to get full claims from a claim cid
-// I'll be honest, I'm not exactly sure whether these claims should be stored or simply synthesized
-// from the information in IPNI combined with having private keys stored in this service
-// I THINK it's possible you can synthesize Index & Equals claims from the information in IPNI + a private key
-// Location commitments are more complicated cause they really ought to be signed by the storer of the commitment?
-type ClaimLookup interface {
-	LookupClaim(model.ProviderResult) (delegation.Delegation, error)
+// ClaimLookup is a read through cache for fetching content claims
+type ClaimLookup struct {
+	httpClient *http.Client
+	claimStore types.ContentClaimsStore
+}
+
+// NewClaimLookup creates a new ClaimLookup with the provided claimstore and HTTP client
+func NewClaimLookup(claimStore types.ContentClaimsStore, httpClient *http.Client) *ClaimLookup {
+	return &ClaimLookup{
+		httpClient: httpClient,
+		claimStore: claimStore,
+	}
+}
+
+// LookupClaim attempts to fetch a claim from either the local cache or via the provided URL (caching the result if its fetched)
+func (cl *ClaimLookup) LookupClaim(ctx context.Context, claimCid cid.Cid, fetchURL url.URL) (delegation.Delegation, error) {
+	// attempt to read claim from cache and return it if succesful
+	claim, err := cl.claimStore.Get(ctx, claimCid)
+	if err == nil {
+		return claim, nil
+	}
+
+	// if an error occurred other than the claim not being in the cache, return it
+	if !errors.Is(err, types.ErrKeyNotFound) {
+		return nil, fmt.Errorf("reading from claim cache: %w", err)
+	}
+
+	// attempt to fetch the claim from provided url
+	resp, err := cl.httpClient.Get(fetchURL.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch claim: %w", err)
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading fetched claim body: %w", err)
+	}
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		return nil, fmt.Errorf("failure response fetching claim. status: %s, message: %s", resp.Status, string(body))
+	}
+	fmt.Println(len(body))
+	claim, err = delegation.Extract(body)
```

Comment:
Who is providing this service? The content claims API currently allows you to fetch claims by CID of the content they're claiming against (the "subject"), not by the CID of the claim (although could be very easily extended to do so).

Reply:
this is just used by the top level service, which allows you get the claim by the content of the cid they're claiming against.

Reply:
the thing to look at really is the content claims top level service.